This phylogeny shows diversity within the Omicron variant from the ongoing novel coronavirus COVID-19 pandemic.
All published Omicron sequences are included, except outliers that differ from main Omicron by more than 10 mutations like hCoV-19/Australia/QLD2584/2021 and SouthAfrica/CERI-KRISP-K032307/2021.

The following sites are masked (plus 100 nt from beginning and 200nt from end): https://github.com/corneliusroemer/ncov-simplest/blob/main/data/exclude_sites_light.txt

These are all the sites that seem to define Omicron and are conserved. Due to the large number of mutations in Spike and deletions, there seem to be amplicon dropouts that are sometimes backfilled with reference. Analysis of trees without masking these revealed a large number of probably artefactual homoplasies.

The analysis is maintained by

* [Cornelius Roemer](https://neherlab.org) at Biozentrum of the University of Basel and the Swiss Institute of Bioinformatics
* [Richard Neher](https://neherlab.org) at Biozentrum of the University of Basel and the Swiss Institute of Bioinformatics

All data we use are deposited in GISAID.

We gratefully acknowledge the authors, originating and submitting laboratories of the genetic sequence and metadata made available through [GISAID](https://gisaid.org) on which this research is based. A full listing of all originating and submitting laboratories is available below. An attribution table is available by clicking on "Download Data" at the bottom of the page and then clicking on "Strain Metadata" in the resulting dialog box.
